api/python/quilt3/packages.py

```diff
@@ -880,10 +914,49 @@ def set(self, logical_key, entry=None, meta=None):
             entry = PackageEntry([url], size, None, orig_meta)
         elif isinstance(entry, PackageEntry):
             entry = entry._clone()
+
+        elif object_is_serializable(entry):
+            # Use file extension from serialization_location, fall back to file extension from logical_key
+            ext = None
+            if len(pathlib.Path(logical_key).suffix) > 0:
+                ext = pathlib.Path(logical_key).suffix[1:]
+            if serialization_location:
+                if len(pathlib.Path(serialization_location).suffix) > 0:
+                    ext = pathlib.Path(serialization_location).suffix[1:]
+
+            format_handlers = FormatRegistry.search(type(entry), ext=ext)
+            if ext:
+                format_handlers = [f for f in format_handlers if ext in f.handled_extensions]
+
+            if len(format_handlers) == 0:
+                error_message = f'Quilt does not know how to serialize a {type(entry)}'
+                error_message_fragment = "."
+                if ext is not None:
+                    error_message_fragment = f' as a {ext} file.'
+                error_message += error_message_fragment + " "
+                error_message += f'If you think this should be supported, please open an issue or PR at ' \
+                                 f'https://github.com/quiltdata/quilt'
+                raise QuiltException(error_message)
+
+            if serialization_format_opts is None:
+                serialization_format_opts = {}
+            serialized_object_bytes, new_meta = format_handlers[0].serialize(entry, meta=None, ext=ext,
+                                                                             **serialization_format_opts)
+            if serialization_location:
+                serialization_path = pathlib.Path(serialization_location).expanduser().absolute()
+                serialization_path.parent.mkdir(exist_ok=True, parents=True)
+            else:
+                suffix = f'.{ext}' if ext else None
+                tmpfile = tempfile.NamedTemporaryFile(delete=False, suffix=suffix)
+                serialization_path = tmpfile.name
+
+            write_url = fix_url(f'file://{serialization_path}')
+            put_bytes(serialized_object_bytes, write_url, meta=new_meta)
+            size, _, _ = get_size_and_meta(write_url)
+            entry = PackageEntry([write_url], size, hash_obj=None, meta=new_meta)
```

Comment:
Why not hash the file?

Reply:
It looks like we leverage parallelism to calculate hashes quickly - https://github.com/quiltdata/quilt/blob/master/api/python/quilt3/data_transfer.py#L703.

If we hash on set, we are removing that opportunity for parallelism.

Reply:
Yes and the set contract is "return right away"

Reply:
Hashing memory-size files isn't a noticeable delay. We could even calculate the hash as we write the bytes to disk, but that's probably over-optimizing.